### PR TITLE
[SPIRV] Prevent creation of jump tables from switch

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR is to prevent creation of jump tables from switch. The reason is that SPIR-V doesn't know how to lower jump tables, and a sequence of commands that IRTranslator generates for switch via jump tables breaks SPIR-V Backend code generation with complains to G_BRJT. The next example is the shortest code to break SPIR-V Backend code generation in this way:

```
target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-n8:16:32:64"
target triple = "spir64-unknown-unknown"

define spir_func void @foo(i32 noundef %val) {
entry:
  switch i32 %val, label %sw.epilog [
    i32 0, label %sw.bb
    i32 1, label %sw.bb2
    i32 2, label %sw.bb3
    i32 3, label %sw.bb4
  ]
sw.bb:
  br label %sw.epilog
sw.bb2:
  br label %sw.epilog
sw.bb3:
  br label %sw.epilog
sw.bb4:
  br label %sw.epilog
sw.epilog:
  ret void
}
```

To resolve the issue we set a high lower limit for number of blocks in a jump table via getMinimumJumpTableEntries() and prevent undesirable (or rather unsupported at the moment) path of code generation.

### CODE REVIEW

File: llvm/lib/Target/SPIRV/SPIRVISelLowering.h [hunk 1]
```diff
@@ -31,6 +31,10 @@ class SPIRVTargetLowering : public TargetLowering {
     return true;
   }
 
+  // Set the lower limit for number of blocks in a jump table in such a way to
+  // prevent creation of jump tables from switch
+  unsigned getMinimumJumpTableEntries() const override { return 0xffffffff; }
```
Comment (Keenuts) [inline llvm/lib/Target/SPIRV/SPIRVISelLowering.h:36]: Wouldn't override `areJTsAllowed` more correct?
https://llvm.org/doxygen/classllvm_1_1TargetLoweringBase.html#a392f6e72d46ff14ee31481e3452f6c31

Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVISelLowering.h]: Thank you, I think you are right, `areJTsAllowed` looks better. Changed

Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVISelLowering.h]: Curious, what happens if the `-fjump-tables` option is used. Do we ignore it?

Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVISelLowering.h]: I think we are forced to manage this on a lower level than user options. This backend clearly doesn't support jump tables at the moment, so lowering switches indeed should ignore -fjump-tables/-fno-jump-tables until the corresponding support is added to the project.

Comment (Keenuts) [reply llvm/lib/Target/SPIRV/SPIRVISelLowering.h]: Sure, wondering if a warning/error should be emitted in the driver if this flag is used with the spirv target.

**Review by Keenuts — APPROVED** [review]

**Review by michalpaszkowski — APPROVED** [review]

File: llvm/lib/Target/SPIRV/SPIRVISelLowering.h (hunk 1 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVISelLowering.h]: It's a good question indeed, thank you. Description of the command line options doesn't help to decide, it's just "Use jump tables for lowering switches" that we can interpret in different ways. Probably, the main argument would be that jump tables provide means of optimization (better code) and at maximum this deserves a warning, not an error. As for warnings, I'm not sure if we produce any at the moment, but all that helps a developer to better understand the job done would help of course. I should remember this idea in general and think how we can use it.

**Pull request merged**